src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoServiceSpec.scala

```diff
@@ -54,5 +103,74 @@ class LeonardoServiceSpec extends TestKit(ActorSystem("leonardotest")) with Flat
+    val googleProject = s"project-${UUID.randomUUID.toString}"
+    
+    // assert that our bucket does not exist
+    assert(!gdDAO.buckets.contains(bucketName))
+
+    // create the bucket and add files
+    leo.initializeBucket(googleProject, clusterName, bucketName)
+
+    // assert that our bucket now exists
+    assert(gdDAO.buckets.contains(bucketName))
+
+    // assert that the init files are in the bucket
+    initFiles.map(initFile => assert(gdDAO.bucketObjects.exists(_ == (bucketName, initFile))))
+  }
+
+  "LeonardoService" should "add bucket objects" in isolatedDbTest {
+    val clusterName = s"cluster-${UUID.randomUUID.toString}"
+    val bucketName = s"bucket-${UUID.randomUUID.toString}"
+    val googleProject = s"project-${UUID.randomUUID.toString}"
+
+    gdDAO.buckets += bucketName
+
+    leo.initializeBucketObjects(googleProject, clusterName, bucketName)
+
+    assert(gdDAO.buckets.contains(bucketName))
+
+    initFiles.map(initFile => assert(gdDAO.bucketObjects.exists(_ == (bucketName, initFile))))
+  }
+
+  "LeonardoService" should "create a firewall rule in a project only once when the first cluster is added" in isolatedDbTest {
+    // set unique names for the google project and the two clusters we'll be creating
+    val cluster1 = s"cluster-${UUID.randomUUID.toString}"
+    val cluster2 = s"cluster-${UUID.randomUUID.toString}"
+    val googleProject = s"project-${UUID.randomUUID.toString}"
+
+    // assert that our google project has no firewall rules
+    assert(!gdDAO.firewallRules.exists(_ == (googleProject, dataprocConfig.clusterFirewallRuleName)))
+
+    // create the first cluster, this should create a firewall rule in our project
+    leo.createCluster(googleProject, cluster1, testClusterRequest).futureValue
+
+    // check that there is exactly 1 firewall rule for our project
+    assert(gdDAO.firewallRules.filterKeys(_ == googleProject).size == 1)
+
+    // create the second cluster. This should check that our project has a firewall rule and not try to add it again
+    leo.createCluster(googleProject, cluster2, testClusterRequest).futureValue
+
+    // check that there is still exactly 1 firewall rule in our project
+    assert(gdDAO.firewallRules.filterKeys(_ == googleProject).size == 1)
+  }
+
+  "LeonardoService" should "template a script using config values" in isolatedDbTest {
+    val clusterName = s"cluster-${UUID.randomUUID.toString}"
+    val bucketName = s"bucket-${UUID.randomUUID.toString}"
+    val googleProject = s"project-${UUID.randomUUID.toString}"
+    val filePath = dataprocConfig.configFolderPath + dataprocConfig.initActionsScriptName
+    val replacements = ClusterInitValues(googleProject, clusterName, bucketName, dataprocConfig).toJson.asJsObject.fields
+
+    val result = leo.template(filePath, replacements).futureValue
+    val expected = "#!/usr/bin/env bash\n\n\"" + clusterName + "\"\n\"" + googleProject + "\"\n\"" + dataprocConfig.jupyterProxyDockerImage + "\""
```

Comment:
This could be a literal string like
```
val expected = 
  s"""
     |!/usr/bin/env bash
     |
     |$clusterName
     |$googleProject
     |${dataprocConfig.jupyterProxyDockerImage}
   """.stripMargin
```